Implement chat messages, mtproto typings and so on

zerobias opened the pull request:
* Implemented mesage slice loading.
Use method api/chats loadSliceRange. It can load various slices ov history, helpful for continious history loading.

* Implemented redux fields chats (possibly should named as dialogs) and users
    *  *chats*. Main dialogs cache, reacts to slice loading, store every dialog messages in normalized form
    *  *users*. Draft realisation. Automatically updates if loaded slice contains unknown users.

* Implemented chat message component. "Dumb" message component also created in Chat folder, because it seems to me that in the future Chat component will become container, and its was better to debug when all child components was located in the same folder.

* Added Mtproto typings. Its really necessary when anybody will add typings to api actions, because almost all api actions begins with server response invoking.

* Moved complex thunk actions out from redux/modules. This dir contains sub-reducers composed to redux state. Possibly it should named
as redux/state or redux/reducers with current reducers.ts as index.ts file.

* Css styling helper classes in helpers/classes

* DC_SERVER env trim

zerobias committed:
Implement chat messages, mtproto typings and so on

* Implemented mesage slice loading.
Use method api/chats loadSliceRange. It can load various slices ov history, helpful for continious history loading.

* Implemented redux fields chats (possibly should named as dialogs) and users
    *  *chats*. Main dialogs cache, reacts to slice loading, store every dialog messages in normalized form
    *  *users*. Draft realisation. Automatically updates if loaded slice contains unknown users.

* Implemented chat message component. "Dumb" message component also created in Chat folder, because it seems to me that in the future Chat component will become container, and its was better to debug when all child components was located in the same folder.

* Added Mtproto typings. Its really necessary when anybody will add typings to api actions, because almost all api actions begins with server response invoking.

* Moved complex thunk actions out from redux/modules. This dir contains sub-reducers composed to redux state. Possibly it should named
as redux/state or redux/reducers with current reducers.ts as index.ts file.

* Css styling helper classes in helpers/classes

* DC_SERVER env trim

goodmind commented on src/app/helpers/Telegram/config.ts:
```diff
@@ -1,9 +1,10 @@
+import { trim } from 'lodash';
```
Change path to `lodash/trim`

goodmind commented on src/app/helpers/Telegram/index.ts:
```diff
@@ -95,8 +97,8 @@ export function generateDialogIndex(date) {
   const promises = Promise.all([
-    invoke('help.getConfig'),
-    invoke('help.getNearestDc'),
+    invoke<IMtpHelpGetConfig>('help.getConfig'),
```
What about something like this? https://goo.gl/JmBEwO

goodmind commented on src/app/redux/api/auth.ts:
```diff
@@ -0,0 +1,86 @@
+import { invoke, APP_HASH, APP_ID, makePasswordHash } from 'helpers/Telegram';
```
Why this in `api/auth` folder? Why just don't leave it in `modules/auth` but rename to `actions.ts` or something like this?

zerobias replied:
api folder is going to be huge, really bigger than modules folder itself. I don`t know where is should be, but now modules folder seems to be pure store implemenation directory

goodmind commented on src/app/redux/modules/messages/index.ts:
```diff
@@ -25,14 +27,14 @@ export const messagesReducer = combineReducers({
+  return (dispatch: IDispatch) => {
     dispatch(MESSAGES.INIT(peer));
-    invoke('messages.getHistory', {
+    invoke<IMtpMessagesSlice>('messages.getHistory', {
```
Again https://goo.gl/JmBEwO

goodmind commented on src/app/redux/modules/users.ts:
```diff
@@ -0,0 +1,33 @@
+    addNewIds(state, users),
+}, []);
+
+// const byId =
```
Where is users?

zerobias committed:
Add tree walk helper

Tree walk filter going to help to sanitize response from server (after invoke).
Remove any function or name with _ excepts _typeName

zerobias committed:
Add users redux store field

zerobias committed:
TreeProcess bugfix

Fixed bug, when function recives array as main object, its return Array-like object

zerobias committed:
Add and improve mtproto schema types

zerobias committed:
Change lodash/trim to ramda at config file

zerobias committed:
Move Message component to separated folder.

Switch to pure function component

zerobias committed:
Bugfix. Remove dispatching fail message on sign in

zerobias committed:
Refactoring. Removed unused models

zerobias committed:
Add state helpers

zerobias committed:
Add store fields

* dialogs

* peers - map id: chat | user | channel

* Loadings - united field for loadings flags

* Selected - united field for selection flags

zerobias committed:
Redux. Rename chats, add new field, users bugfix

Chats store field renmed as histories
New chats field works as dictionary for chat data

zerobias committed:
Refactoring. Remove unused files

zerobias committed:
Rename api/histories as api/chatList

zerobias committed:
Including types from model/auth

zerobias committed:
Enable all implemented redux fields

zerobias committed:
Deep component refactoring.

Separate containers and views. Make view components pure functions

goodmind commented on src/app/components/Message/index.tsx:
```diff
@@ -2,7 +2,7 @@ import * as React from 'react';
 
```
Why not in `Components/Chat/Message`

goodmind commented on src/app/helpers/state.ts:
```diff
@@ -25,6 +41,30 @@ export const appendNew = whenNot(contains, append);
+
+type ITransformer = <F, G>(payload: G) => Array<[number, F]>;
+
+export const changeReducer = <G>(transformer: ITransformer) =>
```
Weird helper...

goodmind commented on src/app/redux/modules/dialogs.ts:
```diff
@@ -0,0 +1,61 @@
+
+const byIdOnDialogsDone = getReduce(getDialogs, addChanged);
+
+const byId = createReducer({
```
`createReducer` have type param which represents state

goodmind commented on src/app/redux/modules/dialogs.ts:
```diff
@@ -0,0 +1,61 @@
+  [GET_DIALOGS.DONE]: byIdOnDialogsDone,
+}, {});
+
+const reducer = combineReducers({
```
Also generic type param

goodmind commented on src/app/redux/modules/loadings.ts:
```diff
@@ -0,0 +1,23 @@
+  chatList: boolean;
+};
+
+const setTrue = () => true;
```
Inconsistent with `modules/auth`, also `ramda` has `R.T` and `R.F`

goodmind commented on src/app/redux/api/histories.ts:
```diff
@@ -0,0 +1,88 @@
+export const loadSliceRange = (dispatch: IDispatch) =>
+  (id: number, peer: IMtpPeer, offset: number = 0, limit: number = 10) => {
+    const adapter = (slice: IMtpMessagesSlice) => {
+      const pure = rejectDashAndFuncs(slice);
```
Why this isn't in `invoke` ?

goodmind commented on src/app/redux/modules/chatList/index.ts:
```diff
@@ -1,161 +0,0 @@
-  return message.from_id;
-};
-
-const saveConversation = (c, m) => dialog => {
```
miss

goodmind commented on src/app/redux/api/chatList.ts:
```diff
@@ -1,14 +1,13 @@
 import { CHATS } from 'actions';
```
Why `api/histories` mixed with `api/dialogs`?

goodmind reviewed (changes requested):
not ok

goodmind committed:
chore(types/ramda): add ramda types

goodmind committed:
build(server): webpack server without server side rendering

goodmind committed:
fix(config): ramda's trim doesn't handle undefined values

goodmind committed:
chore(dev.d.ts): add ramda's reference

goodmind committed:
build(dev): add html-webpack-plugin

goodmind committed:
refactor(histories): add generic types

goodmind committed:
refactor(dialogs): add generic types

goodmind committed:
build(index.html): add html without server side rendering

goodmind committed:
fix(Telegram/Peers): fall-through for channel

goodmind committed:
refactor(Message): change style and template

goodmind committed:
feat(AuthCode): add maxLength prop

goodmind committed:
refactor(treeProcess): add generic types

goodmind committed:
refactor(Message): change style

goodmind committed:
refactor(state): add generic types

goodmind committed:
refactor(peers): add generic types

goodmind committed:
build(package): start script without ssr

goodmind committed:
refactor(Login/PhoneNumber): add dispatch types

goodmind committed:
refactor(Login/Password): add dispatch types

goodmind committed:
fix(ChatListItem): render last message

goodmind committed:
refactor(Home): add dispatch types and change authentication checking

goodmind committed:
chore(gitattributes): fix images

goodmind committed:
feat(Telegram/Files): port of MtpApiFileManager from webogram

goodmind committed:
feat(FileManager): port of FileManager from webogram

goodmind committed:
refactor(PeerPhoto): change mapped state

goodmind committed:
refactor(ChatListItem): import PeerPhoto

goodmind committed:
chore(types/dev): add file APIs types

zerobias committed:
Clean server response inside invoke

(cherry picked from commit 5e674411e5844f2530d81704d14cdd6c58b97bea)

zerobias committed:
Install html-webpack-plugin dev deps

(cherry picked from commit 260498b8d2ee405cca68ed2e3cd38ab94f34e044)

goodmind committed:
feat(FileManager): port from webogram

goodmind committed:
feat(WebpManager): mocked

goodmind committed:
feat(Telegram/Files): downloading small files

goodmind committed:
feat(PeerPhoto): download peer photo

goodmind committed:
refactor(Message): change img to PeerPhoto

zerobias committed:
Separated function for getting connection

(cherry picked from commit b6a106b8f6c9373a51acf5edb361f030c5ed06a9)

zerobias committed:
Merge pull request #2 from goodmind/zerobias-develop

Merge pull request from goodmind 

* feat(FileManager): port from webogram

* feat(WebpManager): mocked

* feat(Telegram/Files): downloading small files

* feat(PeerPhoto): download peer photo

* refactor(Message): change img to PeerPhoto

goodmind committed:
refactor(PeerPhoto): get cached photo

goodmind committed:
refactor(Telegram/Files): uncomment getCachedFile method

zerobias committed:
Replace () => true to ramda equivalent

zerobias committed:
Merge pull request #3 from goodmind/zerobias-develop

* refactor(PeerPhoto): get cached photo
* refactor(Telegram/Files): uncomment getCachedFile method

goodmind commented:
**Roadmap Updates**

- [ ] Mapped MTProto types (https://goo.gl/JmBEwO)
- [x] R.T and R.F

zerobias commented:
R.T and R.F are already implemented at https://github.com/goodmind/treact/pull/8/commits/d4e73154617f49a1b16974b6a0d2ff4a0b5428c5
